### PR TITLE
Feature/school

### LINKED ISSUE #4582: Schoolhouse for Children
I believe that children should all be attending school on a regular basis.  This feature would create a new building called Schoolhouse where children would funnel into each morning to learn, and 1 Teacher will be assigned to run the class.  

A typical day includes:
1. Sit and learn
2. Lunch
3. Recess
4. Sit and learn

By attending school, the children will grow faster into adults.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #4582 

This PR adds a school to the colony.
School give work to one teacher total and up to 2 kids per level.
When kids grow up the automatically leave the school.

Kids only stay in school until noon, then they roam around freely again.

In school kids sit around on carpet and "study". Randomly they get a recess and run around freely in the building.

To compensate kids that study in a school grow up twice as quickly in the remaining time to compensate.

Teachers give out paper to kids and randomly sit down with kids to teach them.

Review please

### CODE REVIEW

**Review by marchermans — DISMISSED** [review]

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/jobs/JobPupil.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java  5
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockHutSchool.java  1
- src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorkerSchool.java  4
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  10
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  9
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutSchool.java  2
- src/main/java/com/minecolonies/coremod/colony/jobs/JobTeacher.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5314839&bid=17394034)

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+     */
+    private static final String TAG_CARPET  = "carpet";
+
+    /**
+     * If the school has a teacher.
+     */
+    private boolean hasTeacher = false;
+
+    /**
+     * Map from beds to patients, 0 is empty.
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java]: so our  childs are patients?

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    @Override
+    public int getMaxInhabitants()
+    {
+        return 1 + 2 * getBuildingLevel();
+    }
+
+    @Override
+    public boolean assignCitizen(final ICitizenData citizen)
+    {
+        if (citizen.isChild() || citizen.getJob() instanceof JobPupil)
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java:130]: do we "reserve" a slot for a teacher?

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        }
+    }
+
+    @Override
+    public CompoundNBT serializeNBT()
+    {
+        final CompoundNBT compound = super.serializeNBT();
+        if (!carpet.isEmpty())
+        {
+            @NotNull final ListNBT bedTagList = new ListNBT();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java]: bedTagList?^^

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+            for (@NotNull final BlockPos pos : carpet)
+            {
+                final CompoundNBT bedCompound = new CompoundNBT();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java]: bedcompound

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+public class EntityAIWorkPupil extends AbstractEntityAIInteract<JobPupil>
+{
+    /**
+     * How often the kid studies for one recess.
+     */
+    private static final int STUDY_TO_RECESS_RATIO = 10;
+
+    /**
+     * To be requested by the teacher.
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java]: from the teacher you mean? else this is in the wrong AI I think

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+          new AITarget(IDLE, START_WORKING, 1),
+          new AITarget(START_WORKING, this::startWorkingAtOwnBuilding, TICKS_SECOND),
+          new AITarget(DECIDE, this::decide, TICKS_SECOND),
+          new AITarget(STUDY, this::study, TICKS_SECOND),
+          new AITarget(RECESS, this::recess, TICKS_SECOND)
+        );
+        worker.setCanPickUpLoot(true);
+    }
+
+    private IAIState decide()
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java:86]: jdoc

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        {
+            worker.getCitizenData().triggerInteraction(new StandardInteractionResponseHandler(new TranslationTextComponent(PUPIL_NO_CARPET), ChatPriority.BLOCKING));
+            return DECIDE;
+        }
+
+        studyPos = pos;
+        return STUDY;
+    }
+
+    private IAIState recess()
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java:110]: jdoc

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        {
+            pupilToTeach = pupils.get(worker.getRandom().nextInt(pupils.size()));
+            return TEACH;
+        }
+
+        return START_WORKING;
+    }
+
+
+    private IAIState teach()
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java:105]: jdoc

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+        if (slot != -1)
+        {
+            InventoryUtils.transferXOfFirstSlotInItemHandlerWithIntoNextFreeSlotInItemHandler(
+              worker.getInventoryCitizen(),
+              PAPER,
+              1, pupilToTeach.getInventoryCitizen()
+            );
+        }
+
+        pupilToTeach.getCitizenData().getCitizenSkillHandler().addXpToSkill(Skill.Intelligence, 1, pupilToTeach.getCitizenData());
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java]: how much is 1 Intelligence xp per second over a half day of school in int levels? not sure if its enough

**Review by someaddons — COMMENTED** [review]
looks good overall, just some doc stuff

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java]: hrm

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java:130]: Good idea.

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java (hunk 9 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java]: added the bonus to the wrong one. (Paper consumption gives big bonus too)

**Comment by reggaemuffin** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java  5
- src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorkerSchool.java  4
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockHutSchool.java  1
- src/main/java/com/minecolonies/coremod/colony/jobs/JobPupil.java  1
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutSchool.java  2
- src/main/java/com/minecolonies/coremod/colony/jobs/JobTeacher.java  1
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  9
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5314839&bid=17394034)

**Comment by marvin-bitterlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkTeacher.java  9
- src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorkerSchool.java  4
- src/main/java/com/minecolonies/coremod/colony/jobs/JobTeacher.java  1
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutSchool.java  2
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockHutSchool.java  1
- src/main/java/com/minecolonies/coremod/colony/jobs/JobPupil.java  1
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/school/EntityAIWorkPupil.java  10
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingSchool.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5314839&bid=17394034)

**Pull request merged**